include/aws/cryptosdk/cache.h

```diff
@@ -531,14 +531,42 @@ AWS_CRYPTOSDK_STATIC_INLINE void aws_cryptosdk_materials_cache_release(
  * @param cache_limit_ttl_units The units of cache_limit_ttl. Allowable values are defined in aws/common/clock.h
  */
 AWS_CRYPTOSDK_API
-struct aws_cryptosdk_cmm *aws_cryptosdk_caching_cmm_new(
+struct aws_cryptosdk_cmm *aws_cryptosdk_caching_cmm_new_from_cmm(
```

Comment:
I find the name `caching_cmm_new_from_cmm` to be slightly confusing, can we be more specific?

Reply:
So C doesn't have classes and objects, but we are essentially simulating a C++ or Java-like behavior here by making functions that are effectively constructors. In C++ or Java, the functions ``aws_cryptosdk_caching_cmm_new_from_cmm`` and ``aws_cryptosdk_caching_cmm_new_from_keyring`` would both be constructors named simply ``caching_cmm``. Then you would allocate a new instance on the heap using the ``new`` keyword with ``new caching_cmm(cmm)`` or ``new caching_cmm(keyring)``, i.e., You would define two different constructors to take either a CMM or keyring as the argument.

However, there are some significant differences in the C language which force us to do things differently.

First, C++ has namespaces and Java has packages, both of which shield the function names from the general namespace and stop our function names from colliding with userspace function names. C has no such thing, so we are making sure that all public names in this SDK start with ``aws_cryptosdk_`` to avoid those collisions.

Second, C has no ``new`` keyword, and what is doing the construction is just a regular function call, so we include the word ``new`` in the function name.

Third, C does not allow you to override the same function name to take arguments of different types the way that C++ and Java do, so you must use different function names in order to take different argument types. For both the session constructors and the caching CMM constructors, the way we are doing this is by using the suffix ``from_cmm`` or ``from_keyring`` to indicate which argument the function takes.

Fourth, when we build up a function name from many constituent parts in this way, it is good practice to put the most general things first and gradually move to the more specific later in the function name. This assures that different functions that are related have the same prefix and makes their grouping more obvious when they are used near each other in the code.

Fifth, both of these functions would only need to take the CMM or the keyring as the argument in C++ or Java, but since C does explicit memory management and aws-c-common has abstracted this into the allocator object, which we need to pass around.

So all of these reasons together are why these construction calls become ``aws_cryptosdk_caching_cmm_new_from_cmm(allocator, cmm)`` and ``aws_cryptosdk_caching_cmm_new_from_keyring(allocator, keyring)``. My opinion is that these names are exactly as specific as they should be, and I don't want to make them any longer than they have already gotten. I hope this explains clearly how we got to these names.

Reply:
Gotcha, thank-you for the explanation. I think I was thinking more on the lines of do we want to explicitly state `upstream cmm`. But on second thought, this seems fine to me.